File: tests/python/test_vpt.py
"""
Python binding tests for VPT (Volume Price Trend).
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestVPT:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_vpt_basic_candles(self, test_data):
        """Test VPT with candle data - mirrors check_vpt_basic_candles"""
        close = test_data['close']
        volume = test_data['volume']
        
        result = ta_indicators.vpt(close, volume)
        assert len(result) == len(close)
    
    def test_vpt_basic_slices(self):
        """Test VPT with basic slice data - mirrors check_vpt_basic_slices"""
        price = np.array([1.0, 1.1, 1.05, 1.2, 1.3])
        volume = np.array([1000.0, 1100.0, 1200.0, 1300.0, 1400.0])
        
        result = ta_indicators.vpt(price, volume)
        assert len(result) == len(price)
        
        
        assert np.isnan(result[0])
        assert np.isnan(result[1])
        
        assert not np.isnan(result[2])
    
    def test_vpt_accuracy_from_csv(self, test_data):
        """Test VPT matches expected values from Rust tests - mirrors check_vpt_accuracy_from_csv"""
        close = test_data['close']
        volume = test_data['volume']
        
        result = ta_indicators.vpt(close, volume)
        
        expected_last_five = [
            -18292.323972247592,
            -18292.510374716476,
            -18292.803266539282,
            -18292.62919783763,
            -18296.152568643138,
        ]
        
        assert len(result) >= 5
        
        
        assert_close(
            result[-5:],
            expected_last_five,
            rtol=0.0,
            atol=1e-9,
            msg="VPT last 5 values mismatch"
        )
    
    def test_vpt_not_enough_data(self):
        """Test VPT fails with insufficient data - mirrors check_vpt_not_enough_data"""
        price = np.array([100.0])
        volume = np.array([500.0])
        
        with pytest.raises(ValueError, match="Not enough valid data"):
            ta_indicators.vpt(price, volume)
    
    def test_vpt_empty_data(self):
        """Test VPT fails with empty data - mirrors check_vpt_empty_data"""
        price = np.array([])
        volume = np.array([])
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.vpt(price, volume)
    
    def test_vpt_all_nan(self):
        """Test VPT fails with all NaN values - mirrors check_vpt_all_nan"""
        price = np.array([np.nan, np.nan, np.nan])
        volume = np.array([np.nan, np.nan, np.nan])
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.vpt(price, volume)
    
    def test_vpt_mismatched_lengths(self):
        """Test VPT fails with mismatched input lengths"""
        price = np.array([1.0, 2.0, 3.0])
        volume = np.array([100.0, 200.0])  
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.vpt(price, volume)
    
    def test_vpt_stream(self):
        """Test VPT streaming calculation"""
        stream = ta_indicators.VptStream()
        
        
        result = stream.update(100.0, 1000.0)
        assert result is None
        
        
        result = stream.update(105.0, 1100.0)
        assert np.isnan(result)
        
        
        result = stream.update(103.0, 1200.0)
        assert not np.isnan(result)
    
    def test_vpt_batch(self, test_data):
        """Test VPT batch processing"""
        close = test_data['close'][:100]  
        volume = test_data['volume'][:100]
        
        result = ta_indicators.vpt_batch(close, volume)
        
        assert 'values' in result
        
        assert 'params' in result
        
        
        values_2d = result['values']
        assert values_2d.shape[0] == 1  
        assert values_2d.shape[1] == len(close)  
        
        
        single_result = ta_indicators.vpt(close, volume)
        assert_close(
            values_2d[0, :], 
            single_result,
            rtol=1e-10,
            msg="Batch vs single VPT mismatch"
        )
    
    def test_vpt_kernel_support(self, test_data):
        """Test VPT with different kernel options"""
        close = test_data['close'][:100]
        volume = test_data['volume'][:100]
        
        
        result_scalar = ta_indicators.vpt(close, volume, kernel='scalar')
        assert len(result_scalar) == len(close)
        
        
        result_auto = ta_indicators.vpt(close, volume)
        assert len(result_auto) == len(close)
        
        
        assert_close(
            result_scalar,
            result_auto,
            rtol=1e-10,
            msg="Kernel results mismatch"
        )
    
    def test_vpt_nan_handling(self, test_data):
        """Test VPT handles NaN values correctly"""
        close = test_data['close']
        volume = test_data['volume']
        
        
        close_with_nan = close.copy()
        volume_with_nan = volume.copy()
        close_with_nan[10] = np.nan
        volume_with_nan[20] = np.nan
        
        result = ta_indicators.vpt(close_with_nan, volume_with_nan)
        assert len(result) == len(close)
        
        
        assert np.isnan(result[0])
        
        
        assert np.isnan(result[11])  
        assert np.isnan(result[21])  

File: tests/wasm/test_vpt.js
/**
 * WASM binding tests for VPT (Volume Price Trend).
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('VPT basic candles', () => {
    
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    
    const result = wasm.vpt_js(close, volume);
    assert.strictEqual(result.length, close.length);
});

test('VPT basic slices', () => {
    
    const price = new Float64Array([1.0, 1.1, 1.05, 1.2, 1.3]);
    const volume = new Float64Array([1000.0, 1100.0, 1200.0, 1300.0, 1400.0]);
    
    const result = wasm.vpt_js(price, volume);
    assert.strictEqual(result.length, price.length);
    
    
    assert(isNaN(result[0]), 'First value should be NaN');
    assert(isNaN(result[1]), 'Second value should be NaN');
    
    assert(!isNaN(result[2]), 'Third value should not be NaN');
});

test('VPT accuracy from CSV', async () => {
    
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    
    const result = wasm.vpt_js(close, volume);
    
    const expected_last_five = [
        -18292.323972247592,
        -18292.510374716476,
        -18292.803266539282,
        -18292.62919783763,
        -18296.152568643138,
    ];
    
    assert(result.length >= 5);
    const last5 = result.slice(-5);
    
    
    assertArrayClose(
        last5,
        expected_last_five,
        1e-9,
        "VPT last 5 values mismatch"
    );
    
    
    await compareWithRust('vpt', result);
});

test('VPT not enough data', () => {
    
    const price = new Float64Array([100.0]);
    const volume = new Float64Array([500.0]);
    
    assert.throws(
        () => wasm.vpt_js(price, volume),
        /Not enough valid data/,
        'Should throw with insufficient data'
    );
});

test('VPT empty data', () => {
    
    const price = new Float64Array([]);
    const volume = new Float64Array([]);
    
    assert.throws(
        () => wasm.vpt_js(price, volume),
        /Empty data/,
        'Should throw with empty data'
    );
});

test('VPT all NaN', () => {
    
    const price = new Float64Array([NaN, NaN, NaN]);
    const volume = new Float64Array([NaN, NaN, NaN]);
    
    assert.throws(
        () => wasm.vpt_js(price, volume),
        /All values are NaN/,
        'Should throw with all NaN values'
    );
});

test('VPT mismatched lengths', () => {
    
    const price = new Float64Array([1.0, 2.0, 3.0]);
    const volume = new Float64Array([100.0, 200.0]); 
    
    assert.throws(
        () => wasm.vpt_js(price, volume),
        /Empty data/,
        'Should throw with mismatched lengths'
    );
});

test('VPT in-place operation', () => {
    
    const price = new Float64Array([1.0, 1.1, 1.05, 1.2, 1.3]);
    const volume = new Float64Array([1000.0, 1100.0, 1200.0, 1300.0, 1400.0]);
    const len = price.length;
    
    
    const out_ptr = wasm.vpt_alloc(len);
    const output = new Float64Array(wasm.__wasm.memory.buffer, out_ptr, len);
    
    
    output.set(price);
    
    
    
    const vol_ptr = wasm.vpt_alloc(len);
    const volumeWasm = new Float64Array(wasm.__wasm.memory.buffer, vol_ptr, len);
    volumeWasm.set(volume);
    
    wasm.vpt_into(out_ptr, vol_ptr, out_ptr, len);
    
    
    wasm.vpt_free(vol_ptr, len);
    
    
    assert(isNaN(output[0]), 'First value should be NaN');
    assert(isNaN(output[1]), 'Second value should be NaN');
    assert(!isNaN(output[2]), 'Should have computed values after in-place operation');
    
    
    wasm.vpt_free(out_ptr, len);
});

test('VPT batch operations', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    const volume = new Float64Array(testData.volume.slice(0, 100));
    
    
    const config = {};
    
    const result = wasm.vpt_batch(close, volume, config);
    
    assert(result.values, 'Should have values array');
    assert.strictEqual(result.rows, 1, 'Should have single row (no parameters)');
    assert.strictEqual(result.cols, close.length, 'Columns should match data length');
    
    
    const single = wasm.vpt_js(close, volume);
    assertArrayClose(
        result.values,
        single,
        1e-10,
        "Batch vs single VPT mismatch"
    );
});

test('VPT NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    
    
    close[10] = NaN;
    volume[20] = NaN;
    
    const result = wasm.vpt_js(close, volume);
    assert.strictEqual(result.length, close.length);
    
    
    assert(isNaN(result[0]), 'First value should be NaN');
    
    
    assert(isNaN(result[11]), 'Value after NaN price should be NaN');
    assert(isNaN(result[21]), 'Value after NaN volume should be NaN');
});
